README.md
# Google Cloud Tasks &middot;  [![NPM](https://img.shields.io/npm/v/google-cloud-tasks.svg?style=flat)](https://www.npmjs.com/package/google-cloud-tasks) [![Tests](https://travis-ci.org/nicolasdao/google-cloud-tasks.svg?branch=master)](https://travis-ci.org/nicolasdao/google-cloud-tasks) [![License](https://img.shields.io/badge/License-BSD%203--Clause-blue.svg)](https://opensource.org/licenses/BSD-3-Clause) [![Neap](https://neap.co/img/made_by_neap.svg)](#this-is-what-we-re-up-to)
__*Google Cloud Tasks*__ is node.js package to push tasks to Google Cloud Tasks (beta). It also include pushing tasks in batches.

# Table of Contents

> * [Install](#install) 
> * [Getting Started](#getting-started)
>	- [Prerequisite](#prerequisite)
>	- [Basics](#basics)
>	- [Four ways to create a client](#four-ways-to-create-a-client)
>		- [User the hosting identity](#user-the-hosting-identity)
>		- [Using a `service-account.json`](#using-a-service-accountjson)
>		- [Using explicit credentials](#using-explicit-credentials)
>		- [Using environment variables](#using-environment-variables)
>	- [How To Test Locally](#how-to-test-locally)
>	- [Other Utilities](#other-utilities)
>	- [Minimizing Network Errors](#minimizing-network-errors)
> * [About Neap](#this-is-what-we-re-up-to)
> * [Annexes](#annexes)
>	- [Available Regions](#available-regions)
> * [License](#license)


# Install
```
npm i google-cloud-tasks --save
```

# Getting Started

## Prerequisite

Before using this package, you must first:

1. Have a Google Cloud Account.
2. Have a Project in that Google Account (the next step are specific to that Project). __WARNING__: As of today (June 2019), Google Cloud Tasks API is in beta. That means that not all locations are available. Make sure that your App Engine is running in one of the regions described in the [Annexes](#annexes)/[Available Regions](#available-regions).
> Google keeps adding more location until this service moves from beta to GA. You can double the latest list at [https://cloud.google.com/tasks/docs/dual-overview](https://cloud.google.com/tasks/docs/dual-overview). 
3. Have an App Engine service running.
4. Have a Task Queue configured to push tasks to the App Engine service above.
5. Have a Service Account or a hosting service associated with a service account set up with the following 3 roles :
	- `roles/appengine.appViewer`
	- `roles/cloudtasks.enqueuer`
	- `roles/cloudtasks.viewer`

If hosting environment is Cloud Compute, App Engine, Cloud Function or Cloud Run, and if their associated service account has all the roles above, then that's it. If, on the other hand, you wish to use a service account to explicitly create a new client, then:

1. Get the JSON keys file for that Service Account above
2. Save that JSON key into a `service-account.json` file. Make sure it is located under a path that is accessible to your app (the root folder usually).
3. Add a `location_id` property into that `service-account.json` file. That property should contain the location of your App Engine. Because the Google Cloud Tasks API is currently in beta, only the regions described in the [Annexes](#annexes)/[Available Regions](#available-regions) are available.

## Basics

```js
const { join } = require('path')
const { client } = require('google-cloud-tasks')

// There is only one queue per App Engine service. If you have multiple App Engine microservices,
// you have to create a queue per service.
const queue = client.new({
	name: 'your-queue-name', // Required. This is the Google Cloud Task that points to a specific App Engine Service.
	method: 'POST',		 // Optional. Default 'GET'
	headers: {		 // Optional. Default {}
		Accept: 'application/json'
	},
	jsonKeyFile: join(__dirname, './service-account.json')	// Required
})

const task_01 = {  
	name: 'task #1', 
	otherData: {
		age: 23
	} 
}

const task_02 = {  
	name: 'task #2', 
	otherData: {
		age: 32
	} 
}

// Sending a single task to queue 'your-queue-name' using a 'POST' at the pathname 'pathname-01'
// with headers 'Accept: application/json'
queue.task('pathname-01').send(task_01)
	.then(({ status, data }) => console.log({ status, data }))

// Sending multiple tasks to queue 'your-queue-name' using a 'POST' at the pathname 'pathname-01'
// with headers 'Accept: application/json'
queue.task('pathname-01').send([task_01, task_02])
	.then(({ status, data, errors }) => console.log({ status, data, errors })) 	// data is an response array for each successfull task sent
																				// errors is an response array for each failed task

// Sending a single task to queue 'your-queue-name' using a 'POST' at the pathname '/'
// with headers 'Accept: application/json'
queue.task().send(task_01)
	.then(({ status, data }) => console.log({ status, data }))

// Sending a single task to queue 'your-queue-name' using a 'POST' at the pathname 'pathname-01'
// with headers 'Accept: application/json' and 'Authorization: 123'
queue.task('pathname-01', { headers: { Authorization: '123' } }).send(task_01)
	.then(({ status, data }) => console.log({ status, data }))

// Sending a single task to queue 'your-queue-name' using a 'POST' at the pathname 'pathname-01'
// with headers 'Accept: application/json', 'Authorization: 123' and 'Custom: some other data'
queue.task('pathname-01', { headers: { Authorization: '123' } }).send(task_01, { headers: { Custom: 'some other data' } })
	.then(({ status, data }) => console.log({ status, data }))

// Delaying to send a single task to queue 'your-queue-name' using a 'POST' at the pathname 'pathname-01'
// with headers 'Accept: application/json' and 'Custom: some other data' to the 1st of Feb 2020
queue.task('pathname-01').send(task_01, { schedule: new Date(2020,1,1), headers: { Custom: 'some other data' }})
	.then(({ status, data }) => console.log({ status, data }))

// Preventing to send the same task more than once to queue 'your-queue-name' using a 'POST' at the pathname 'pathname-01'
// with headers 'Accept: application/json' and 'Custom: some other data'. This is done by explicitely settin the task id. 
// That id must be unique. The second task will fail because the task with id 1 will already been added.
queue.task('pathname-01').send(task_01, { id:1, headers: { Custom: 'some other data' }})
	.then(({ status, data }) => console.log({ status, data }))
queue.task('pathname-01').send(task_01, { id:1, headers: { Custom: 'some other data' }})
	.then(({ status, data }) => console.log({ status, data }))

// Sending a single task to queue 'your-queue-name' using a 'POST' at the pathname 'pathname-01'
// with headers 'Accept: application/json' and 'Custom: some other data'. Explicitely setting the task id based on the task payload.
// NOTICE: 	In the example, the second argument of the 'send' method is not an object anymore, but a function. This function 
// 			is supposed to return an object.
queue.task('pathname-01').send(task_01, t => ({ id: t.otherData.age, headers: { Custom: 'some other data' } }))
	.then(({ status, data }) => console.log({ status, data }))
```

## Four ways to create a client

This library supports four different ways to create a client. The first method is the recommended way:
1. [User the hosting identity](#user-the-hosting-identity)
2. [Using a `service-account.json`](#using-a-service-accountjson)
3. [Using explicit credentials](#using-explicit-credentials)
4. [Using environment variables](#using-environment-variables)

### User the hosting identity

```js
const { client } = require('google-cloud-tasks')

// There is only one queue per App Engine service. If you have multiple App Engine microservices,
// you have to create a queue per service.
const queue = client.new({
	name: 'your-queue-name', // Required. This is the Google Cloud Task that points to a specific App Engine Service.
	method: 'POST',		 // Optional. Default 'GET'
	headers: {		 // Optional. Default {}
		Accept: 'application/json'
	},
	locationId: 'australia-southeast1'	// Required
})
```

In this case, the package fetches the credentials automatically. It will try three different techniques to get those data, and if none of them work, an error is thrown. Those techniques are:
1. If the code is hosted on GCP (e.g., Cloud Compute, App Engine, Cloud Function or Cloud Run) then the credentials are extracted from the service account associated with the GCP service.
2. If the `GOOGLE_APPLICATION_CREDENTIALS` environment variable exists, its value is supposed to be the path to a service account JSON key file on the hosting machine.
3. If the `~/.config/gcloud/application_default_credentials.json` file exists, then the credentials it contains are used (more about setting that file up below).

When developing on your local environment, use either #2 or #3. #3 is equivalent to being invited by the SysAdmin to the project and granted specific privileges. To set up `~/.config/gcloud/application_default_credentials.json`, follow those steps:

- Make sure you have a Google account that can access both the GCP project and the resources you need on GCP. 
- Install the `GCloud CLI` on your environment.
- Execute the following commands:
	```
	gcloud auth login
	gcloud config set project <YOUR_GCP_PROJECT_HERE>
	gcloud auth application-default login
	```
	The first command logs you in. The second command sets the `<YOUR_GCP_PROJECT_HERE>` as your default project. Finally, the third command creates a new `~/.config/gcloud/application_default_credentials.json` file with the credentials you need for the `<YOUR_GCP_PROJECT_HERE>` project.

### Using a `service-account.json`

We assume that you have created a Service Account in your Google Cloud Account (using IAM) and that you've downloaded a `service-account.json` (the name of the file does not matter as long as it is a valid json file). The first way to create a client is to provide the path to that `service-account.json` as shown in the following example:

```js
const { join } = require('path')
const { client } = require('google-cloud-tasks')

// There is only one queue per App Engine service. If you have multiple App Engine microservices,
// you have to create a queue per service.
const queue = client.new({
	name: 'your-queue-name', // Required. This is the Google Cloud Task that points to a specific App Engine Service.
	method: 'POST',		 // Optional. Default 'GET'
	headers: {		 // Optional. Default {}
		Accept: 'application/json'
	},
	jsonKeyFile: join(__dirname, './service-account.json')	// Required
})
```

### Using explicit credentials

This method is similar to the previous one. You should have dowloaded a `service-account.json`, but instead of providing its path, you provide some of its details explicitly:

```js
const { client } = require('google-cloud-tasks')

// There is only one queue per App Engine service. If you have multiple App Engine microservices,
// you have to create a queue per service.
const queue = client.new({
	name: 'your-queue-name', // Required. This is the Google Cloud Task that points to a specific App Engine Service.
	method: 'POST',		 // Optional. Default 'GET'
	headers: {		 // Optional. Default {}
		Accept: 'application/json'
	},
	credentials: {
		client_email:'something-1234@your-project-id.iam.gserviceaccount.com', 
		private_key: '-----BEGIN PRIVATE KEY-----\n123456789-----END PRIVATE KEY-----\n'
	}
})
```

### Using environment variables

```js
const { client } = require('google-cloud-tasks')

// There is only one queue per App Engine service. If you have multiple App Engine microservices,
// you have to create a queue per service.
const queue = client.new({
	name: 'your-queue-name', // Required. This is the Google Cloud Task that points to a specific App Engine Service.
	method: 'POST',		 // Optional. Default 'GET'
	headers: {		 // Optional. Default {}
		Accept: 'application/json'
	}
})
```

The above will only work if all the following environment variables are set:
- `GOOGLE_CLOUD_TASK_PROJECT_ID` or `GOOGLE_CLOUD_PROJECT_ID`
- `GOOGLE_CLOUD_TASK_REGION` or `GOOGLE_CLOUD_REGION`
- `GOOGLE_CLOUD_TASK_CLIENT_EMAIL` or `GOOGLE_CLOUD_CLIENT_EMAIL`
- `GOOGLE_CLOUD_TASK_PRIVATE_KEY` or `GOOGLE_CLOUD_PRIVATE_KEY`

> WARNING: If you're using [NPM's `dotenv`](https://www.npmjs.com/package/dotenv), wrap your PRIVATE_KEY between double-quotes, otherwise some characters are escaped which corrupts the key.

Refer to the next section to see how to pass an OAuth2 token.


## How To Test Locally

Because the task queue is nothing but a queue sitting in front of a real web server, there is very little difference between sending messages to the queue or directly to the underlying HTTP endpoint. The only real difference is that hitting the underlying HTTP endpoint will be immediate instead of being scheduled by the queue. Being able to bypass the queue is usually useful during the design or testing phase. Here is an example on how you can send messages directly to the underlying HTTP endpoint:

```js
const queue = client.new({
	name: 'your-queue-name',								// Required
	jsonKeyFile: join(__dirname, './service-account.json'),
	byPassConfig: {
		service: 'http://localhost:4000'
	}
})

const task_01 = {  
	name: 'task #1', 
	otherData: {
		age: 23
	} 
}

// Sending a single task using a 'POST' to 'http://localhost:4000/pathname-01' bypassing the queue 'your-queue-name'.
queue.task('pathname-01').send(task_01)
	.then(({ status, data }) => console.log({ status, data }))
```

## Other Utilities
### Listing all the pending tasks
> WARNING: Only scan up to 30,000 tasks. If there are more, these APIs will return incomplete results.

List all tasks for that queue
```js
queue.task().list().then(tasks => console.log(tasks))
```

Only list the tasks for the specific pathname 'pathname-01' in that queue
```js
queue.task('pathname-01').list().then(tasks => console.log(tasks))
```

### Finding a specific task
> WARNING: Only scan up to 30,000 tasks. If there are more, these APIs will return incomplete results.

Return the first task that matches the task predicate (a task contains the following property: `id`, `method`, `pathname`, `schedule` and `created`)
```js
queue.task().find(({ id }) => /^123/.test(id)).then(task => console.log(task))
```

Does the same as above, but only for tasks with a pathname equal to 'pathname-01'
```js
queue.task('pathname-01').find(({ id }) => /^123/.test(id)).then(task => console.log(task))
```

### Testing if a task exists
> WARNING: Only scan up to 30,000 tasks. If there are more, these APIs will return incomplete results.

Return the first task that matches the task predicate (a task contains the following property: `id`, `method`, `pathname`, `schedule` and `created`)
```js
queue.task().some(({ id }) => /^123/.test(id)).then(yes => console.log(yes))
```

Does the same as above, but only for tasks with a pathname equal to 'pathname-01'
```js
queue.task('pathname-01').some(({ id }) => /^123/.test(id)).then(yes => console.log(yes))
```

### Testing if an http request is from Cloud Task Or CRON

```js
const { utils: { isTaskRequest, isCronRequest } } = require('google-cloud-tasks')

app.get('/', (req,res) => {
	const message = 
		isTaskRequest(req) ? `I'm a Cloud Task request` : 
		isCronRequest(req) ? `I'm a CRON request` : `I'm a standard request` 

	res.status(200).send(message)
})
```

> Note: Both `isTaskRequest` and `isCronRequest` accept either a request (e.g., { headers: { ... } }) object or a headers object (e.g., { ... }).
The way those 2 methods work is straighforward. They look for the existence of those 2 headers in the request: `'x-appengine-queuename'` and `'x-appengine-taskname'`. If those 2 headers are both present, the request is either a CRON or a Cloud Task request. A CRON request is one with a `'x-appengine-queuename'`	 equal to `'__cron'`.

### Formatting a Task ID

A Task ID is an optional argument that can be passed when pushing a task to the queue. It helps to make sure that the task is only pushed once. That task ID only accept the following characters: a-z, A-Z, 0-9, - and _ 

To facilitate generating those IDs, we provide the following utility:

```js
const { utils: { formatTaskId } } = require('google-cloud-tasks')

const taskId = formatTaskId(`You can type anything here %&532vj% @V~`)
```

## Minimizing Network Errors

Networks errors (e.g. socket hang up, connect ECONNREFUSED) are a fact of life. To deal with those undeterministic errors, this library uses a simple exponential back off retry strategy, which will reprocess your read or write request for 10 seconds by default. You can increase that retry period as follow:

```js
queue.task('pathname-01').send(task_01, { timeout: 30000 })

queue.task().list({ timeout: 30000 })

queue.task('pathname-01').find(({ id }) => /^123/.test(id), { timeout: 30000 })

queue.task('pathname-01').some(({ id }) => /^123/.test(id), { timeout: 30000 })
```

# This Is What We re Up To
We are Neap, an Australian Technology consultancy powering the startup ecosystem in Sydney. We simply love building Tech and also meeting new people, so don't hesitate to connect with us at [https://neap.co](https://neap.co).

Our other open-sourced projects:
#### GraphQL
* [__*graphql-serverless*__](https://github.com/nicolasdao/graphql-serverless): GraphQL (incl. a GraphiQL interface) middleware for [webfunc](https://github.com/nicolasdao/webfunc).

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.0.0](https://github.com/nicolasdao/google-cloud-tasks/compare/v0.6.1...v1.0.0) (2020-08-20)


### Features

* Add support for application default credentials ([12ad89f](https://github.com/nicolasdao/google-cloud-tasks/commit/12ad89f05970d92c564cc45a6912e50fa5aff751))

### [0.6.1](https://github.com/nicolasdao/google-cloud-tasks/compare/v0.6.0...v0.6.1) (2020-07-16)


### Bug Fixes

* Vulnerability ([34f142a](https://github.com/nicolasdao/google-cloud-tasks/commit/34f142a5f2791a744fc2021dc932506cec5c67f2))

<a name="0.6.0"></a>
# [0.6.0](https://github.com/nicolasdao/google-cloud-tasks/compare/v0.5.6...v0.6.0) (2019-02-09)


### Features

* Add support for sanitizing task ID ([e0ae03f](https://github.com/nicolasdao/google-cloud-tasks/commit/e0ae03f))



<a name="0.5.6"></a>
## [0.5.6](https://github.com/nicolasdao/google-cloud-tasks/compare/v0.5.5...v0.5.6) (2019-01-25)


### Bug Fixes

* A single failing task prevent all the subsequent others to not be pushed. Dev mode does not support failing is the same task id is pushed more than once ([256d2d0](https://github.com/nicolasdao/google-cloud-tasks/commit/256d2d0))



<a name="0.5.5"></a>
## [0.5.5](https://github.com/nicolasdao/google-cloud-tasks/compare/v0.5.4...v0.5.5) (2019-01-23)


### Bug Fixes

* The Dev mode does not emit request with the 'x-appengine-queuename' and 'x-appengine-taskname' headers, preventing to properly identify those requests as coming from the Task Queue ([48e1daa](https://github.com/nicolasdao/google-cloud-tasks/commit/48e1daa))



<a name="0.5.4"></a>
## [0.5.4](https://github.com/nicolasdao/google-cloud-tasks/compare/v0.5.3...v0.5.4) (2019-01-18)


### Bug Fixes

* options should never be null ([1f2a65c](https://github.com/nicolasdao/google-cloud-tasks/commit/1f2a65c))



<a name="0.5.3"></a>
## [0.5.3](https://github.com/nicolasdao/google-cloud-tasks/compare/v0.5.2...v0.5.3) (2019-01-18)


### Bug Fixes

* Undefined indentity ref in the utils lib ([957bfb8](https://github.com/nicolasdao/google-cloud-tasks/commit/957bfb8))



<a name="0.5.2"></a>
## [0.5.2](https://github.com/nicolasdao/google-cloud-tasks/compare/v0.5.1...v0.5.2) (2019-01-17)


### Features

* Make pushing a task in dev mode not exceed 4 secs ([515bcda](https://github.com/nicolasdao/google-cloud-tasks/commit/515bcda))



<a name="0.5.1"></a>
## [0.5.1](https://github.com/nicolasdao/google-cloud-tasks/compare/v0.5.0...v0.5.1) (2018-12-12)


### Features

* Add support for scheduling task in dev mode ([aeb026f](https://github.com/nicolasdao/google-cloud-tasks/commit/aeb026f))



<a name="0.5.0"></a>
# [0.5.0](https://github.com/nicolasdao/google-cloud-tasks/compare/v0.4.1...v0.5.0) (2018-12-10)


### Features

* Add more robust retry mechanism ([9694a59](https://github.com/nicolasdao/google-cloud-tasks/commit/9694a59))



<a name="0.4.1"></a>
## [0.4.1](https://github.com/nicolasdao/google-cloud-tasks/compare/v0.4.0...v0.4.1) (2018-12-09)


### Features

* Add a random delay when pushing tasks in Dev mode in order to avoid stressing the system too much ([b0cb2c2](https://github.com/nicolasdao/google-cloud-tasks/commit/b0cb2c2))



<a name="0.4.0"></a>
# [0.4.0](https://github.com/nicolasdao/google-cloud-tasks/compare/v0.3.7...v0.4.0) (2018-12-06)


### Features

* Add support for listing pending tasks and testing if pending tasks exists ([d85c5d4](https://github.com/nicolasdao/google-cloud-tasks/commit/d85c5d4))



<a name="0.3.7"></a>
## [0.3.7](https://github.com/nicolasdao/google-cloud-tasks/compare/v0.3.6...v0.3.7) (2018-12-05)



<a name="0.3.6"></a>
## [0.3.6](https://github.com/nicolasdao/google-cloud-tasks/compare/v0.3.5...v0.3.6) (2018-12-05)


### Features

* Add support to detect whether an HTTP request is from a CRON job or a Task Queue ([0230353](https://github.com/nicolasdao/google-cloud-tasks/commit/0230353))



<a name="0.3.5"></a>
## [0.3.5](https://github.com/nicolasdao/google-cloud-tasks/compare/v0.3.4...v0.3.5) (2018-11-25)


### Bug Fixes

* Fetching non-JSON data fails, which forces the retry mechanism to kick off unecessarilly ([7136a44](https://github.com/nicolasdao/google-cloud-tasks/commit/7136a44))



<a name="0.3.4"></a>
## [0.3.4](https://github.com/nicolasdao/google-cloud-tasks/compare/v0.3.3...v0.3.4) (2018-11-22)


### Bug Fixes

* unit test is failing because of wrong UTC format ([d4ea953](https://github.com/nicolasdao/google-cloud-tasks/commit/d4ea953))



<a name="0.3.3"></a>
## [0.3.3](https://github.com/nicolasdao/google-cloud-tasks/compare/v0.3.2...v0.3.3) (2018-11-22)


### Bug Fixes

* unit test is failing because of wrong UTC format ([71dd97d](https://github.com/nicolasdao/google-cloud-tasks/commit/71dd97d))



<a name="0.3.2"></a>
## [0.3.2](https://github.com/nicolasdao/google-cloud-tasks/compare/v0.3.1...v0.3.2) (2018-11-18)


### Bug Fixes

* Validate url throws an exception when using a valid localhost ([6a17954](https://github.com/nicolasdao/google-cloud-tasks/commit/6a17954))



<a name="0.3.1"></a>
## [0.3.1](https://github.com/nicolasdao/google-cloud-tasks/compare/v0.3.0...v0.3.1) (2018-11-18)


### Features

* Add support for bypassing the task queue and sending the message straight to the underlying service ([b384783](https://github.com/nicolasdao/google-cloud-tasks/commit/b384783))



<a name="0.3.0"></a>
# [0.3.0](https://github.com/nicolasdao/google-cloud-tasks/compare/v0.2.1...v0.3.0) (2018-11-18)


### Features

* Add support for new API style ([2cf08df](https://github.com/nicolasdao/google-cloud-tasks/commit/2cf08df))



<a name="0.2.1"></a>
## [0.2.1](https://github.com/nicolasdao/google-cloud-tasks/compare/v0.2.0...v0.2.1) (2018-11-15)



<a name="0.2.0"></a>
# [0.2.0](https://github.com/nicolasdao/google-cloud-tasks/compare/v0.1.0...v0.2.0) (2018-11-15)


### Features


index.js
/**
 * Copyright (c) 2018, Neap Pty Ltd.
 * All rights reserved.
 * 
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
*/

const { GoogleAuth } = require('google-auth-library')
const { promise: { retry }, collection, obj: { merge }, validate } = require('./utils')
const { pushTask, listTasks, findTask } = require('./src/gcp')

const ERR_INVALID_SCHEDULE_CODE = 589195462987

const getToken = auth => auth.getAccessToken()

const _retryFn = (fn, options={}) => retry(
	fn, 
	() => true, 
	err => {
		const throwErrorNow = err && err.message && err.message.toLowerCase().indexOf('lacks iam permission') >= 0
		return !throwErrorNow
	},
	{ ignoreFailure: true, retryInterval: [500, 2000], timeOut: options.timeout || 10000 })


const _getJsonKey = jsonKeyFile => require(jsonKeyFile)

const HTTP_METHODS = { 'GET': true, 'POST': true, 'PUT': true, 'DELETE': true, 'PATCH': true, 'OPTIONS': true, 'HEAD': true }
/**
 * Creates a new Google Cloud Task client. 
 *
 * @param  {String} config.name 					
 * @param  {String} config.method 					
 * @param  {String} config.pathname 					
 * @param  {String} config.headers 					
 * @param  {String} config.jsonKeyFile 					Path to the service-account.json file. If specified, 'clientEmail', 'privateKey', 'projectId' are not required.
 * @param  {String} config.credentials.project_id
 * @param  {String} config.credentials.client_email
 * @param  {String} config.credentials.private_key
 * @param  {String} config.projectId
 * @param  {String} config.locationId 					
 * @return {Object}        				
 */
const createClient = config => {
	let { name, method, pathname, headers={}, jsonKeyFile, credentials, mockFn, byPassConfig={}, projectId, locationId } = config || {}
	const { getJsonKey: _mockGetJsonKey, pushTask: _mockPushTask, getToken: _mockGetToken } = mockFn || {}
	let { project_id, location_id, client_email, private_key } = credentials ? credentials : jsonKeyFile ? (_mockGetJsonKey || _getJsonKey)(jsonKeyFile) : {}

	projectId = projectId || project_id || process.env.GOOGLE_CLOUD_TASK_PROJECT_ID || process.env.GOOGLE_CLOUD_PROJECT_ID
	locationId = locationId || location_id || process.env.GOOGLE_CLOUD_TASK_REGION || process.env.GOOGLE_CLOUD_REGION
	client_email = client_email || process.env.GOOGLE_CLOUD_TASK_CLIENT_EMAIL || process.env.GOOGLE_CLOUD_CLIENT_EMAIL
	private_key = private_key || process.env.GOOGLE_CLOUD_TASK_PRIVATE_KEY || process.env.GOOGLE_CLOUD_PRIVATE_KEY

	const { service:serviceUrl } = byPassConfig
	if (serviceUrl && !validate.url(serviceUrl))
		throw new Error(`Invalid argument exception. 'byPassConfig.service' ${serviceUrl} is an invalid URL.`)

	const getTheToken = serviceUrl ? (() => Promise.resolve('dev-token-1234')) : (_mockGetToken || getToken)
	const pushTheTask = _mockPushTask || pushTask

	const queue = name
	if (!queue)
		throw new Error('Missing required argument \'name\'')

	method = (method || 'GET').trim().toUpperCase() 
	pathname = pathname || '/'
	
	if (!HTTP_METHODS[method])
		throw new Error(`Invalid argument exception. Method '${method}' is not a valid HTTP method.`)
	
	if (!locationId)
		throw new Error(`Missing required 'locationId' property. This property should be explicitly defined or should be present inside the service account json file ${jsonKeyFile}. It contains the location where the Google Cloud Task Queue is hosted.`)
	
	const authConfig = {
		scopes: ['https://www.googleapis.com/auth/cloud-platform']
	}

	if (client_email && private_key)
		authConfig.credentials = { client_email, private_key }

	const auth = new GoogleAuth(authConfig)

	const getProjectId = () => Promise.resolve(null)
		.then(() => projectId ? projectId : auth.getProjectId())
		.then(id => {
			if (!id)
				throw new Error(`Missing required 'projectId' property. This property should be explicitly defined or should be present inside the service account json file ${jsonKeyFile}.`)
			if (!projectId)
				projectId = id 

			return id
		})

	const push = (taskData, options={}) => getProjectId().then(_projectId => { 
		const { id, method:_method, schedule, pathname:_pathname, headers:_header={}, body={} } = taskData
		const h = merge(headers, _header)
		let s
		if (schedule) {
			const scheduleType = typeof(schedule)
			if (schedule instanceof Date)
				s = schedule.toISOString()
			else if (scheduleType == 'string') {
				const d = new Date(schedule)
				if (d.toString().toLowerCase() == 'invalid date') {
					let e = new Error(`Invalid argument exception. 'schedule' ${schedule} is an invalid date`)
					e.code = ERR_INVALID_SCHEDULE_CODE
					throw e
				}
				s = d.toISOString()
			} else if (scheduleType == 'number') {
				const ref = Date.now() + 5000
				const n = ref > schedule ? ref : schedule
				const d = new Date(n)
				if (d.toString().toLowerCase() == 'invalid date') {
					let e = new Error(`Invalid argument exception. 'schedule' ${schedule} is an invalid date`)
					e.code = ERR_INVALID_SCHEDULE_CODE
					throw e
				}
				s = d.toISOString()
			} else {
				let e = new Error(`Invalid argument exception. 'schedule' ${schedule} is an invalid date`)
				e.code = ERR_INVALID_SCHEDULE_CODE
				throw e
			}
		}
		return getTheToken(auth)
			.then(token => _retryFn(
				() => pushTheTask({
					id,
					schedule: s,
					projectId: _projectId, 
					locationId, 
					queueName: queue, 
					token, 
					method: _method || method,
					pathname: _pathname || pathname, 
					headers: h,
					body,
					serviceUrl
				}),
				options))
	})
		.catch(e => {
			if (options.retryCatch)
				return options.retryCatch(e)
			else
				throw e
		})
		.then(({ status, data, request }) => {
			if (status >= 200 && status < 300)
				return data
			else {
				let message = request && request.method && request.uri 
					? `Pushing task to queue '${name}' using an HTTP ${request.method} to ${request.uri} failed with HTTP code ${status}`
					: `Pushing task to queue '${name}' failed with HTTP code ${status}`

				if (data && data.error && data.error.message)
					message = `${message}\nDetails: ${data.error.message}`
			
				let e = new Error(message)
				e.data = data || {}
				e.status = status // 409 means that a task with the ID already exists
				throw e
			}
		})

	const service = {
		/**
		 * [description]
		 * @param  {Object}  taskData 					Task payload
		 * @param  {Number}  options.retryCatch 		If specified, this function will deal with managing exception in case
		 *                                        		the retry attempts all fail. If this option is not specified,  
		 * @type {[type]}
		 */
		push,
		/**
		 * [description]
		 * @param  {Array}   batchData 					Array of task payload
		 * @param  {Number}  options.batchSize 			Default is 200
		 * @param  {Number}  options.timeout 			Default 10,000. Used to configure the length of the retry strategy
		 * @param  {Number}  options.retryCatch 		If sepcified, this function will deal with managing exception in case
		 *                                        		the retry attempts all fail. If this option is not specified, a single failure
		 *                                        		will cause the interruption of the entire batch  
		 * @param  {Boolean} options.debug  			Default is false. Shows details if set to true         
		 * @return {[type]} 							[description]
		 */
		batch: (batchData, options={}) => Promise.resolve(null).then(() => {
			options = options || {}
			if (!batchData)
				return { status: 200, data: {} }
			if (!Array.isArray(batchData))
				throw new Error('Wrong argument exception. \'batchData\' must be an array.')

			const batchSize = options.batchSize && options.batchSize > 0 ? options.batchSize : 200
			return collection.batch(batchData, batchSize).reduce((runJob, taskDataBatch) => 
				runJob.then(taskResponses => {
					const start = Date.now()
					return Promise.all(taskDataBatch.map(t => push(t, options).catch(err => ({ id: t.id, error: { status: err.status, message: err.message } }))))
						.then(values => {
							if (options.debug)
								console.log(`Batch with ${taskDataBatch.length} tasks has been enqueued in ${((Date.now() - start)/1000).toFixed(2)} seconds`)
							return values.reduce((acc, data) => {
								if (data)
									acc.push(data)
								return acc
							}, taskResponses)
						})
				}), Promise.resolve([]))
				.then(responses => {
					responses = responses || []
					const failedTasks = responses.filter(x => x.error)
					const failedCount = failedTasks.length
					if (failedCount > 0) {
						let e = new Error(`${failedCount} task${failedCount > 1 ? 's' : ''} failed and ${batchData.length - failedCount} succeeded (out of ${batchData.length} pushed tasks)`)
						e.data = failedTasks
						throw e
					} else
						return responses
				})
		})
	}

	return {
		push: service.push,
		batch: service.batch,
		task: (pathName, options={}) => {
			if (pathName && typeof(pathName) == 'string') 
				pathName = pathName.trim()
			const pathNameIsOptions = typeof(pathName) == 'object'
			const { method:_method, headers:_headers } = pathNameIsOptions ? pathName : (options || {})
			const p = pathNameIsOptions ? pathname : (pathName || pathname)
			const filter = pathName ? (({ pathname }) => pathname == `/${pathName.replace(/(^\/*|\/*$)/g, '')}`) : null

			const find = (fn, options={}) => getTheToken(auth)
				.then(token => getProjectId().then(id => ({ projectId:id, token })))
				.then(({ token, projectId:_projectId }) => {
					if (!fn)
						throw new Error('Missing required \'fn\' argument.')
					if (typeof(fn) != 'function')
						throw new Error(`Invalid argument exception. 'fn' must be a function (current: ${typeof(fn)}).`)

					return _retryFn(
						() => findTask({ projectId: _projectId, locationId, queueName: queue, token, find: fn }, { filter }),
						options)
				})

			return {
				list: (options={}) => getTheToken(auth)
					.then(token => getProjectId().then(id => ({ projectId:id, token })))
					.then(({ token, projectId:_projectId }) => _retryFn(
						() => listTasks({ projectId: _projectId, locationId, queueName: queue, token }, { filter }), 
						options)),
				find,
				some: (fn, options={}) => find(fn, options).then(result => result ? true : false),
				send: (task, options={}) => Promise.resolve(null).then(() => {
					const optionsType = typeof(options)
					const optionsIsFn = optionsType == 'function'
					if (optionsType != 'object' && !optionsIsFn)
						throw new Error('Invalid argument exception. \'options\' must either be an object or a function returning an object.')

					const optionsFn = optionsIsFn ? options : () => options
					const getParams = t => {
						const { id, headers:__headers, schedule } = optionsFn(t) || {} 
						const h = merge(headers, _headers, __headers)
						return { id, schedule, headers: h }
					}
					
					if (Array.isArray(task)) {
						let firstOptions = null
						const batchTasks = task.map(t => {
							if (!firstOptions)
								firstOptions = optionsFn(t)
							const { id, schedule, headers } = getParams(t)
							return {
								id, 
								method: _method || method, 
								schedule, 
								pathname: p, 
								headers, 
								body: t
							}
						})
						return service.batch(batchTasks, firstOptions)
					} else {
						const firstOptions = optionsFn(task)
						const { id, schedule, headers } = getParams(task)
						return service.push({
							id, 
							method: _method || method, 
							schedule, 
							pathname: p, 
							headers, 
							body: task
						}, firstOptions)
					}
				})
			}
		}
	}
}

const isHeadersFromCloudTask = headers => {
	headers = headers || {}
	const queueName = headers['x-appengine-queuename']
	return headers['x-appengine-taskname'] && (queueName && queueName != '__cron') ? true : false
}

const isHeadersFromCron = headers => {
	headers = headers || {}
	const queueName = headers['x-appengine-queuename']
	return headers['x-appengine-taskname'] && (queueName && queueName == '__cron') ? true : false
}

const isRequestFromCloudTask = req => {
	req = req || {}
	return isHeadersFromCloudTask(req.headers) || isHeadersFromCloudTask(req)
}

const isRequestFromCron = req => {
	req = req || {}
	return isHeadersFromCron(req.headers) || isHeadersFromCron(req)
}

const formatTaskId = id => encodeURIComponent(id || '').replace(/[^a-zA-Z0-9-_]/g, '')

module.exports = {
	client: {
		new: createClient
	},
	utils: {
		isTaskRequest: isRequestFromCloudTask,
		isCronRequest: isRequestFromCron,
		formatTaskId
	}
}



